src/python/pants/core/target_types.py

```diff
@@ -42,19 +57,114 @@
 from pants.util.docutil import bin_name
 from pants.util.frozendict import FrozenDict
 from pants.util.logging import LogLevel
+from pants.util.meta import frozen_after_init
 from pants.util.strutil import softwrap
 
 # -----------------------------------------------------------------------------------------------
-# `file` and`files` targets
+# Asset target helpers
 # -----------------------------------------------------------------------------------------------
 
 
-class FileSourceField(SingleSourceField):
+@dataclass(unsafe_hash=True)
+@frozen_after_init
+class HTTPSource:
+    url: str
+    len: int
+    sha256: str
+    # Defaults to last part of the URL path (E.g. `index.html`)
+    filename: str
+
+    def __init__(self, url: str, *, len: int, sha256: str, filename: str = ""):
+        for field in dataclasses.fields(self):
+            value = locals()[field.name]
+            if not isinstance(value, getattr(builtins, cast(str, field.type))):
+                raise TypeError(f"`{field.name}` must be a `{field.type}`, got `{type(value)!r}`.")
+
+        self.url = url
+        self.len = len
+        self.sha256 = sha256
+        self.filename = filename or self.url.rsplit("/", 1)[1]
+
+        if not self.filename:
+            raise ValueError(
+                "Couldn't deduce filename from `url`. Please specify the `filename` argument."
+            )
+        if "\\" in self.filename or "/" in self.filename:
+            raise ValueError("`filename` cannot contain a path separator.")
```

Comment:
, but was set to `{}`

Reply:
We need file+line exceptions. Then this wouldn't need to be so parrot-y